### PR TITLE
Librefomo

### PR DESCRIPTION
**Pull request by bencejdanko**
Issue: https://github.com/LibreYOLO/libreyolo/issues/300

LibreFOMO has been integrated using point-based task, predicting [x, y, class, confidence].

Currently, it supports loading random initialization and pre-trained imagenet weights. 

For random initialization, set model_path=None:

```
model = LibreFOMO(model_path=None, size=model_size, nb_classes=nb_classes, device='0')
```

For ImageNet pretrained weights, it loads from this source: https://huggingface.co/fomo-edge-ai/FOMO/tree/main/weights (cc-by-nc-4.0, a license warning has also been added)

To load, set model_path="LibreFOMOm.pt", for example:

```
model = LibreFOMO(model_path="LibreFOMOm.pt", size=model_size, nb_classes=nb_classes, device='0')
```

Using ImageNet pretrained measurably improves results. 
E.g w/ random weight initialization, we get ~0.8714 F1 (the grid). With ImageNet-pretrained, we get ~0.8952 F1. With Edge Impulse studio, we got ~0.92 F1 on the same dataset (garlic on a conveyer belt).
The demonstration of training LibreFOMO using this PR branch is in this notebook: https://github.com/fomo-edge-ai/fomo/blob/master/examples/libreyolo_librefomo_example.ipynb

image

Some design decisions:

- We override the PointValidator with FOMOValidator(PointValidator) such that we log not just the initial standard point metrics, but the FOMO-specific ones for scoring the grid heatmap. We save based off of the best F1 off the heatmap instead of the usual "fitness" (mAP sweep for points).
- The weights were converted from tensorflow to a PyTorch checkpoint using this script: https://github.com/fomo-edge-ai/fomo/blob/master/scripts/convert_keras_mobilenetv2_to_pytorch.py. By default names={0: "person"}, but the script detects the classes from the `data.yaml` configuration and replaces the head on initialization.
- Technically multiclass is supported, but each grid can only be occupied by a single point estimation.

## Release Notes

* **New Features**
  * Added the **LibreFOMO** point-localization model to the public API, including pretrained small/medium/large weights and point grid decoding for inference.
  * Introduced FOMO-specific training and validation (custom preprocessing, loss, dataset utilities, and configurable confidence/NMS-distance settings).

* **Compliance**
  * Added a **one-time LibreFOMO license notice** before downloading compatible weights.

* **Tests**
  * Added end-to-end and unit test coverage for model loading, inference, training, datasets, validation preprocessing, decoding, and postprocessing.

### CODE REVIEW

File: libreyolo/models/fomo/model.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    @classmethod
+    def detect_size(cls, weights_dict: dict) -> Optional[str]:
+        return detect_size_from_state_dict(weights_dict)
+
+    @classmethod
+    def detect_nb_classes(cls, weights_dict: dict) -> Optional[int]:
+        weight = weights_dict.get("head.weight")
+        if weight is None:
+            return None
+        return max(int(weight.shape[0]) - 1, 1)
```
[the remark this answers is not in the document]
Comment (bencejdanko) [reply libreyolo/models/fomo/model.py:69]: This issue appears widespread (lack of DDP support)

**Comment by bencejdanko** [comment]
Updated based on reviews:

- Updated the checkpoint aliases to use `LibreFOMOs-point.pt` with `-point` suffix
- Updated nomenclature.md to reflect FOMO and point usage
- Updated the non-commercial license warning log to appear immediately on checkpoint download, similar to YOLO-NAS
- Added `FOMOTrainTransform`, for when using augmentations

Updated sample training run at https://github.com/fomo-edge-ai/fomo/blob/master/examples/libreyolo_librefomo_example.ipynb

**Comment by EHxuban11** [comment]
Hello @bencejdanko , this is looking really nice, I have trained some models to test and it's working. I need more time to review. I'll try to finish the review today.

**Comment by EHxuban11** [comment]
I trained them on the aerial sheep dataset, dog dataset and strawberrry dataset and it works reasonably well. I literally can't believe that it has only around 20k parameters. I'm in shock.

aerial-sheep_val_02
aerial-sheep_val_03
aerial-sheep_val_04
dog_val_01
strawberry_test_04

**Comment by EHxuban11** [comment]
Hello @bencejdanko , code is fine from my side. I also tried training models and it worked great: https://github.com/LibreYOLO/libreyolo/pull/403#issuecomment-4772719743

Fenomenal job, this is amazing!
